From 42050133dad52c56f1f27864e537b9ebc12d5e45 Mon Sep 17 00:00:00 2001
From: Kaifeng Zhu <kaifeng.zhu@citrix.com>
Date: Fri, 7 Mar 2014 09:50:41 +0000
Subject: [PATCH 01/10] readline: fix memory corruption when adding history.
To: andrew.cooper3@citrix.com

idx can be down to 0, so TERM_MAX_CMDS-idx+1 could be TERM_MAX_CMDS+1, which
exceeds the size of term_history.

Coverity issue #21185.

Signed-off-by: Kaifeng Zhu <kaifeng.zhu@citrix.com>
diff --git a/readline.c b/readline.c
index 4294af9..e08b459 100644
--- a/readline.c
+++ b/readline.c
@@ -268,7 +268,7 @@ static void term_hist_add(const char *cmdline)
 	    new_entry = hist_entry;
 	    /* Put this entry at the end of history */
 	    memmove(&term_history[idx], &term_history[idx + 1],
-		    (TERM_MAX_CMDS - idx + 1) * sizeof(char *));
+		    (TERM_MAX_CMDS - (idx + 1)) * sizeof(char *));
 	    term_history[TERM_MAX_CMDS - 1] = NULL;
 	    for (; idx < TERM_MAX_CMDS; idx++) {
 		if (term_history[idx] == NULL)
